processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java

```diff
@@ -327,7 +394,34 @@ public Cursor apply(final Long input)
                   {
                     private final Offset initOffset = offset.clone();
                     private final DateTime myBucket = gran.toDateTime(input);
-                    private Offset cursorOffset = offset;
+                    private final CursorOffsetHolder cursorOffsetHolder = new CursorOffsetHolder();
+                    {
+                      cursorOffsetHolder.set(offset);
+                    }
+
+                    private final CursorOffsetHolderValueMatcherFactory valueMatcherFactory;
+                    private final RowOffsetMatcherFactory rowOffsetMatcherFactory;
+                    {
+                      valueMatcherFactory = new CursorOffsetHolderValueMatcherFactory(cursorOffsetHolder, this);
```

Comment:
Or, only create them if postFilter is non-empty. It looks like they're only used there, and are not needed otherwise.